### ISSUE TITLE
Rule proposal: checks if you remembered to call super.method()

### ISSUE THREAD

**Issue by trusktr** [typescript-eslint/typescript-eslint#684]
**Expected Result**

A rule that can detects if we forget to call a super method. Most of the time I want to call super methods. It is possible to forget to call a super method, and it can cause subtle bugs that aren't discoverable until later (depending on the implementation of the method of course).

**Actual Result**

There's not a rule that I know of for this yet.

**Additional Info**

Originally posted at https://github.com/eslint/eslint/issues/11969, but then we realized that there's no type info, so for plain JS ESLint might not be able to know if a super method exists in order to verify if you are calling super.

**Comment by djmisterjon** [comment]
it will hard to know if is required or not ?
maybe if ts can add new field like `require` on a method, so if we override a not `final` method, rule can understand we needed to call super.

**Comment by mahdi-farnia** [comment]
Now typescript added support for `override` keyword and a compilerOption `noImplicitOverride` to force adding `override` keyword.

Not any update on this issue?

**Comment by bradzacher** [comment]
With any issue in opened in this project - it either has a visible progress in the form of an attached PR, or it has no progress.

We are a community run project. The **volunteer** maintainers spend most of their time triaging issues and reviewing PRs. This means that most issues *will not progress* unless a member of the community steps up and **champions** it.

If this issue is important to you - consider being that champion.

If not - please just subscribe to the issue and wait patiently.
Commenting asking for status updates does not bump issue priority in any way and just serves to spam ***everyone*** subscribed to the issue.

**Comment by mahdi-farnia** [comment]
> With any issue in opened in this project - it either has a visible progress in the form of an attached PR, or it has no progress.
> 
> We are a community run project. The **volunteer** maintainers spend most of their time triaging issues and reviewing PRs. This means that most issues _will not progress_ unless a member of the community steps up and **champions** it.
> 
> If this issue is important to you - consider being that champion.
> 
> If not - please just subscribe to the issue and wait patiently. Commenting asking for status updates does not bump issue priority in any way and just serves to spam _**everyone**_ subscribed to the issue.

I would like to be the champion :)
I'm not too expert in OSS, I've contributed in some projects with little changes, but I will do my best! 😁

**Comment by mahdi-farnia** [comment]
> 👋 just checking in @mahdi-farnia, is this still something you have time for? And if so, anything we can do to help? 

Hello, and thank you for that.


Actually yes, but I'm busy for my college works and it goes really slow. If you are working on that, i think it's better to have your code instead of mine. ( I am new to AST & something like this )


I have done the main part and documenting but i stuck at testing.


The code that checks wether you call super.method or not is not working as expected. That's the problem I'm facing with.

**Comment by JoshuaKGoldberg** [comment]
Oh great! Tell you what: if you can send a draft PR, I would love to build off what you have. That way we can give you credit in the repository's history for the work you've already done. 😄

**Comment by JoshuaKGoldberg** [comment]
This sounds perfect. I love the strategy - great idea! 👏

**Comment by mahdi-farnia** [comment]
> This sounds perfect. I love the strategy - great idea! 👏

https://github.com/typescript-eslint/typescript-eslint/compare/main...mahdi-farnia:typescript-eslint:main

Hope it can solve the problem

**Comment by bradzacher** [comment]
Thinking about this - what about the cases where you are explicitly providing a new implementation without calling the base method - is that just a "use a disable comment" case?

**Comment by bradzacher** [comment]
I'm still mulling this over from the POV of looking at the cases that you don't want to call the overridden method.

One example - it's entirely valid (and some would probably say *recommended*) to mark an abstract implementation as `override`:
```ts
abstract class Foo {
  abstract method(): string;
}

class Bar extends Foo {
  override method() { return 'Bar' }
}
```
[playground](https://www.typescriptlang.org/play?noImplicitOverride=true#code/PTAEAEDsHsEkFsAOAbAlgY1QFwPIDcBTAJyNQBMCAoSgQwCMBnLIm9LUdZGhh0AMWjRQAb0qhxoekxZtJjZqywBZAlgAW0MgEYAFAEoAXKGmpIAcwDcYiVIWzbM5ao1kATPqMnzVgL7VO3LwAQjREoAQAHlgEkGS8AkKiEqAgoDDRoAC0aULERNBhAGb58DkIKBjY+HnkBKCFBaDqqLzo3HXcTWrEBADkvDAc0JDoRKp1qEjIBPAxWDRYqMOgdASmZqDQhCTkFJDW4g6KKuqaunoiB8mgY1gArkSQoL0hRL1WyX7JV1s1FHLSY7OTTuC5Ja7iW4PJ4vULvK5+PxAA)

In this case it's a TS error to call `super.method()` because TS knows that it has no implementation to call.
Side note that `noImplicitOverride` will not flag this case it if is missing the `override` keyword - that compiler option only flags overrides of concrete implementations.

Which means that this lint rule will *have* to use types - because suggesting to the user they should call the super method that provably doesn't exist is quite non-sensical!
We could concede this case and document that you need to use a disable comment - but it seems like a bit of a hole in the rule.

There's also what I mentioned before - you are going to have cases where you're intentionally not calling the override as you're providing an alternate implementation rather than an extended one.
Unlike before - this one does seem like it's probably "okay" to document with a disable comment so that others can clearly see you've intentionally not called the super.

I would flag the [case of a declaration merged interface](https://www.typescriptlang.org/play?noImplicitOverride=true#code/PTAEAEDsHsEkFsAOAbAlgY1QFwPIDcBTAJyNQBMCAoS1SLYgMwEN0DQAxaaUAb3gKwALaGQAUASgBcoAM5ZSkAOYBfSumRMZMjl16q1GraABCTIqAIAPepDLbO3HpVAvQ0QiXJt+QkRN7OrkEyAK6IxAB0PsJi4gDcoCCgAO6oyMig6ESagqBMWKBEIXSo-IFBLkQCIUSQoADk9XHloKrKQA), though this code is already SUPER dangerous and dumb, so it's not something I expect us to handle ([`no-unsafe-declaration-merging` bans this](https://typescript-eslint.io/rules/no-unsafe-declaration-merging/))

I don't do enough class-based programming to know if there are other cases that I haven't thought of where it's valid to not call the super method.

If someone has some more insight into OOP in TS it'd be appreciated - more insight would aid in trying to figure out which path forward we need to take:
a) if this rule is fine as a pure AST check,
b) if this rule needs type info, or
c) if this rule would have too many false-positives to be a rule in this project.

**Comment by mahdi-farnia** [comment]
I agree with you, it's necessary for checking in such cases like abstract classes!

I'm a little busy nowadays, I’ll work on this as fast as i can.
Sorry for late response 😅

**Comment by Josh-Cena** [comment]
> if this rule is fine as a pure AST check,

In that case, the rule would not be very useful in a codebase without `noImplicitOverrides`. That may be an okay thing to do, but we have to make it clear in the docs.

> if this rule would have too many false-positives to be a rule in this project.

That depends _a lot_ on the project's code style. I would say it encourages a quite strange style, because method shadowing is quite common. For example, none of the `toString()` method overrides would ever call `super.toString()`, because that produces `"[object Object]"`. However, maybe for those who opt into this rule, they choose to enforce this kind of style.

I would advise against implementing it as a core rule. Not until someone has played with it in their own plugin.

**Comment by JoshuaKGoldberg** [comment]
Yeah I'm not positive enough about the rule to advocate for it being here. +1 to this:

> Note for readers: that's not to say it's not a good rule - just that it doesn't fit into our general-purpose plugin.

@mahdi-farnia since you already started on #5848: I'd recommend making it its own standalone package that folks can pull in if they want. If you're still working on it and want help, ping me and I can try to be useful. 🙂

**Comment by mahdi-farnia** [comment]
I agree
